File: pkg/controller/client/channel.go
package client

import (
	"context"
	"net/http"

	"inspr.dev/inspr/pkg/api/models"
	"inspr.dev/inspr/pkg/meta"
	"inspr.dev/inspr/pkg/meta/utils/diff"
	"inspr.dev/inspr/pkg/rest"
	"inspr.dev/inspr/pkg/rest/request"
)

// ChannelClient is a client for manipulating Channel structures in Insprd
type ChannelClient struct {
	reqClient *request.Client
}

// Get gets a channel from Insprd
// The scope refers to the parent app of the given channel, represented with a dot separated query
// such as app1.app2. The name is the name of the channel.
func (cc *ChannelClient) Get(ctx context.Context, scope, name string) (*meta.Channel, error) {
	cdi := models.ChannelQueryDI{
		ChName: name,
	}
	var resp meta.Channel

	err := cc.reqClient.
		Header(rest.HeaderScopeKey, scope).
		Send(ctx, "/channels", http.MethodGet, cdi, &resp)
	if err != nil {
		return nil, err
	}

	return &resp, nil
}

// Create creates given channel inside of Insprd
// The scope refers to the parent app of the given channel, represented with a dot separated query
// such as app1.app2
func (cc *ChannelClient) Create(ctx context.Context, scope string, ch *meta.Channel, dryRun bool) (diff.Changelog, error) {
	cdi := models.ChannelDI{
		Channel: *ch,
		DryRun:  dryRun,
	}
	var resp diff.Changelog

	err := cc.reqClient.
		Header(rest.HeaderScopeKey, scope).
		Send(ctx, "/channels", http.MethodPost, cdi, &resp)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

// Delete deletes a channel inside the Insprd
// The scope refers to the parent app of the given channel, represented with a dot separated query
// such as app1.app2. The name is the name of the channel to be deleted
func (cc *ChannelClient) Delete(ctx context.Context, scope, name string, dryRun bool) (diff.Changelog, error) {
	cdi := models.ChannelQueryDI{
		ChName: name,
		DryRun: dryRun,
	}
	var resp diff.Changelog

	err := cc.reqClient.
		Header(rest.HeaderScopeKey, scope).
		Send(ctx, "/channels", http.MethodDelete, cdi, &resp)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

// Update updates given channel structure, if it exists in Insprd
// The scope refers to the parent app of the given channel, represented with a dot separated query
// such as app1.app2
func (cc *ChannelClient) Update(ctx context.Context, scope string, ch *meta.Channel, dryRun bool) (diff.Changelog, error) {
	cdi := models.ChannelDI{
		Channel: *ch,
		DryRun:  dryRun,
	}
	var resp diff.Changelog

	err := cc.reqClient.
		Header(rest.HeaderScopeKey, scope).
		Send(ctx, "/channels", http.MethodPut, cdi, &resp)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

File: clients/python/inspr/controller/channel.py
import sys
from ..models import *
from ..rest import *
from .client import *

CHANNEL_ROUTE = "channels"

class ChannelClient(Client):
    def get(self, scope:str, channel_name:str) -> InsprStructure:
        msg_body = {
            "chname": channel_name
        }
        
        headers = self.get_header_with_scope(scope)

        try:
            resp = send_get_request(self.url + "/" + CHANNEL_ROUTE, body=msg_body, headers=headers)
            return InsprStructure(json.loads(resp.text))
        except Exception as e:
            raise Exception(f"Error while send a Get Channel request: {e}")

    def delele(self, scope:str, channel_name:str, dryRun:bool) -> Changelog:
        msg_body = {
            "chname": channel_name,
            "dry": dryRun
        }

        headers = self.get_header_with_scope(scope)

        try:
            resp = send_delete_request(self.url + "/" + CHANNEL_ROUTE, body=msg_body, headers=headers)
            return Changelog(json.loads(resp.text))
        except Exception as e:
            raise Exception(f"Error while send a Delete Channel request: {e}")

    def create(self, scope:str, channel:dict, dryRun:bool) -> Changelog:
        msg_body = {
            "channel": channel,
            "dry": dryRun
        }

        headers = self.get_header_with_scope(scope)

        try:
            resp = send_post_request(self.url + "/" + CHANNEL_ROUTE, msg_body, headers)
            return Changelog(json.loads(resp.text))
        except Exception as e:
            raise Exception(f"Error while send a Create Channel request: {e}")

    def update(self, scope:str, channel:dict, dryRun:bool) -> Changelog:
        msg_body = {
            "channel": channel,
            "dry": dryRun
        }

        headers = self.get_header_with_scope(scope)

        try:
            resp = send_update_request(self.url + "/" + CHANNEL_ROUTE, msg_body, headers)
            return Changelog(json.loads(resp.text))
        except Exception as e:
            raise Exception(f"Error while send a Update Channel request: {e}")